File: src/ba03/ba03b_str_spell_by_path/ba03b.cpp
/*
Rosalind: BA3B
String Spelled by a Genome Path Problem

Find the string spelled by a genome path.

Given: A sequence of k-mers Pattern[1], ... , Pattern[n]
such that the last k - 1 symbols of Pattern[i] are equal
to the first k - 1 symbols of Pattern[i+1] for i from 1 to n-1.

Return: A string Text of length k+n-1 where the i-th k-mer in Text is equal to Patterni for all i.

Sample Dataset
ACCGA
CCGAA
CGAAG
GAAGC
AAGCT

Sample Output
ACCGAAGCT

═════════════════════════════════════════════════

    [ Where am I? ]

    * How do we assemble genome?
      ↓
    * String reconstruction problem
      -> PREV: Generate the k-mer Composition of a String (BA3A)
      -> HERE: String Reconstruction Problem with k-mer composition(BA3B)
      ↓
    * NEXT: Construct OverapGraph (BA3C)

References:
- Best way to extract a subvector from a vector?
  https://stackoverflow.com/questions/421573/best-way-to-extract-a-subvector-from-a-vector
*/

#include <unordered_set>
#include <limits>
#include <random>
#include <time.h>
#include "/home/wsl/rosalind/include/utils.hpp"

using namespace std;

// BA03B: string spell by ordered genome path
string str_spell(vector<string> patterns) {
    string result = patterns[0];
    vector<string> subvec{ patterns.begin() + 1, patterns.end() };
    for (auto& pattern : subvec) {
        result += pattern.substr(pattern.length() - 1, 1);
    }
    return result;
}

// BA03B: string spell by ordered genome path
string str_spell2(vector<string> patterns) {
    string result = patterns[0];
    vector<string> subvec{ patterns.begin() + 1, patterns.end() };
    for (auto it = patterns.begin() + 1; it != patterns.end(); it++) {
        result += (*it).substr(it->length() - 1, 1);  // two different ways of dereferencing
    }
    return result;
}

// main
int main() {
    vector<string> lines = read_lines("/home/wsl/rosalind/data/ba03b.txt");

    clock_t tStart = clock();
    cout << str_spell2(lines) << endl;
    printf("Time taken: %.6fs\n", (double)(clock() - tStart) / CLOCKS_PER_SEC);
}

File: src/ba03/ba03b_str_spell_by_path/go/ba03b.go
/*
Rosalind: BA3B
String Spelled by a Genome Path Problem

Find the string spelled by a genome path.

Given: A sequence of k-mers Pattern[1], ... , Pattern[n]
such that the last k - 1 symbols of Pattern[i] are equal
to the first k - 1 symbols of Pattern[i+1] for i from 1 to n-1.

Return: A string Text of length k+n-1 where the i-th k-mer in Text is equal to Patterni for all i.

Sample Dataset
ACCGA
CCGAA
CGAAG
GAAGC
AAGCT

Sample Output
ACCGAAGCT

═════════════════════════════════════════════════

    [ Where am I? ]

    * How do we assemble genome?
      ↓
    * String reconstruction problem
      -> PREV: Generate the k-mer Composition of a String (BA3A)
      -> HERE: String Reconstruction Problem with k-mer composition(BA3B)
      ↓
    * NEXT: Construct OverapGraph (BA3C)
*/

package main

import (
	"bufio"
	"fmt"
	"os"
	"time"
)

// BA03B: string spelled by genome path
func str_spelled_by_genome_path(patterns []string) string {
	result := patterns[0]
	for _, s := range patterns[1:] {
		result += string(s[len(s)-1])
	}
	return result
}

// main
func main() {
	lines := ReadLines("/home/wsl/rosalind/data/ba03b.txt")

	start := time.Now()
	fmt.Println(str_spelled_by_genome_path(lines))
	fmt.Println("Execution time:", time.Since(start))
}

// helper fx: read lines
func ReadLines(path string) []string {
	file, err := os.Open(path)
	if err != nil {
		return nil
	}
	defer file.Close()

	var lines []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}
	return lines
}
